fabric-component-access-api-v1/src/main/java/net/fabricmc/fabric/impl/component/access/ComponentTypeImpl.java

```diff
@@ -0,0 +1,218 @@
+	private final IdentityHashMap<Item, Function<ItemComponentContext, ?>> itemMappings = new IdentityHashMap<>();
+	private final IdentityHashMap<EntityType<?>, Function<EntityComponentContext, ?>> entityMappings = new IdentityHashMap<>();
+	private final IdentityHashMap<Item, ObjectArrayList<BiPredicate<ItemComponentContext, T>>> itemActions = new IdentityHashMap<>();
+
+	private ObjectArrayList<Pair<Function<EntityComponentContext, T>, Predicate<EntityType<?>>>> deferedEntityMappings;
+
+	private final AbsentComponentAccess<T> absentComponentAccess;
+
+	ComponentTypeImpl(T absent) {
+		this.absent = absent;
+		defaultBlockMapping = b -> absent;
+		defaultItemMapping = i -> absent;
+		defaultEntityMapping = e -> absent;
+		absentComponentAccess = new AbsentComponentAccess<>(this);
+	}
+
+	@Override
+	public T absent() {
+		return absent;
+	}
+
+	Function<BlockComponentContext, ?> getMapping(Block block) {
+		return blockMappings.getOrDefault(block, defaultBlockMapping);
+	}
+
+	@Override
+	public void registerBlockProvider(Function<BlockComponentContext, T> mapping, Block... blocks) {
+		for (final Block b : blocks) {
+			blockMappings.put(b, mapping);
+		}
+	}
+
+	Function<EntityComponentContext, ?> getMapping(Entity entity) {
+		return entityMappings.getOrDefault(entity.getType(), defaultEntityMapping);
+	}
+
+	@Override
+	public void registerEntityProvider(Function<EntityComponentContext, T> mapping, EntityType<?>... entities) {
+		for (final EntityType<?> e : entities) {
+			entityMappings.put(e, mapping);
+		}
+	}
+
+	Function<ItemComponentContext, ?> getMapping(Item item) {
+		return itemMappings.getOrDefault(item, defaultItemMapping);
+	}
+
+	@Override
+	public void registerItemProvider(Function<ItemComponentContext, T> mapping, Item... items) {
+		for (final Item i : items) {
+			itemMappings.put(i, mapping);
+		}
+	}
+
+	@Override
+	public void registerItemAction(BiPredicate<ItemComponentContext, T> action, Item... items) {
+		for (final Item i : items) {
+			ObjectArrayList<BiPredicate<ItemComponentContext, T>> list = itemActions.get(i);
+
+			if (list == null) {
+				list = new ObjectArrayList<>(4);
+				itemActions.put(i, list);
+			}
+
+			list.add(action);
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public ComponentAccess<T> getAccess(World world, BlockPos pos) {
+		return BlockComponentContextImpl.get(this, world, pos);
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public ComponentAccess<T> getAccess(World world, BlockPos pos, BlockState blockState) {
+		return BlockComponentContextImpl.get(this, world, pos, blockState);
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public ComponentAccess<T> getAccess(BlockEntity blockEntity) {
+		return BlockComponentContextImpl.get(this, blockEntity);
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public <E extends Entity> ComponentAccess<T> getAccess(E entity) {
+		applyDeferredEntityRegistrations();
```

Comment:
This is a bit fragile, mods could create dummy entities during init.

Reply:
Just to confirm, you mean they may call this during init with a transient Entity instance before registration is final.  Is that right?

Does Fabric yet have an event or other supported means to defer this until registries are final? I'd have to look.

Reply:
That's indeed what I thought about. I don't know if there is a better way that doesn't involve entrypoints though.